extensions-core/druid-basic-security/src/main/java/org/apache/druid/security/basic/authentication/BasicAuthenticatorUserPrincipal.java

```diff
@@ -0,0 +1,137 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.security.basic.authentication;
+
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.security.basic.BasicAuthUtils;
+import org.apache.druid.security.basic.authentication.entity.BasicAuthenticatorCredentials;
+
+import javax.naming.ldap.LdapName;
+import java.security.Principal;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+public class BasicAuthenticatorUserPrincipal implements Principal
+{
+  private static final Logger LOG = new Logger(BasicAuthenticatorUserPrincipal.class);
+
+  private final String name;
+  private final BasicAuthenticatorCredentials credentials;
+  private final Set<LdapName> groups;
+  private final Instant createdAt;
+  private final AtomicReference<Instant> lastVerified = new AtomicReference<>();
+
+  public BasicAuthenticatorUserPrincipal(
+      String name,
+      BasicAuthenticatorCredentials credentials,
+      Set<LdapName> groups
+  )
+  {
+    this(name, credentials, groups, Instant.now());
+  }
+
+  private BasicAuthenticatorUserPrincipal(
+      String name,
+      BasicAuthenticatorCredentials credentials,
+      Set<LdapName> groups,
+      Instant createdAt
+  )
+  {
+    Objects.requireNonNull(name, "name is required");
+    Objects.requireNonNull(credentials, "credentials is required");
+    Objects.requireNonNull(groups, "groups is required");
+    Objects.requireNonNull(createdAt, "createdAt is required");
+
+    this.name = name;
+    this.credentials = credentials;
+    this.groups = groups;
+    this.createdAt = createdAt;
+    this.lastVerified.set(createdAt);
+  }
+
+  @Override
+  public String getName()
+  {
+    return this.name;
+  }
+
+  public Set<LdapName> getGroups()
+  {
+    return groups;
+  }
+
+  public Instant getCreatedAt()
+  {
+    return createdAt;
+  }
+
+  public Instant getLastVerified()
+  {
+    return lastVerified.get();
+  }
+
+  public boolean hasSameCredentials(char[] password)
+  {
+    byte[] recalculatedHash = BasicAuthUtils.hashPassword(
+        password,
+        this.credentials.getSalt(),
+        this.credentials.getIterations()
+    );
+    if (Arrays.equals(recalculatedHash, credentials.getHash())) {
+      this.lastVerified.set(Instant.now());
+      LOG.debug("Refereshing lastVerified principal user '%s'", this.name);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  public boolean isExpired(int duration, int maxDuration)
+  {
+    long now = System.currentTimeMillis();
+    long cutoff = now - (duration * 1000L);
+    if (this.lastVerified.get().toEpochMilli() < cutoff) {
```

Comment:
From the docs, the `maxDuration` expiring should override the result of any last verified checks, the maxCutoff check should always happen and only if it passes should last verified time be checked

Reply:
Fixed